adapters/adocean/adocean.go

```diff
@@ -0,0 +1,317 @@
+	if consentString != "" {
+		queryParams.Add("gdpr_consent", consentString)
+		queryParams.Add("gdpr", "1")
+	}
+	if request.User != nil && request.User.BuyerUID != "" {
+		queryParams.Add("hcuserid", request.User.BuyerUID)
+	}
+	endpointURL.RawQuery = queryParams.Encode()
+
+	return endpointURL.String(), nil
+}
+
+func validateParams(params *openrtb_ext.ExtImpAdOcean) error {
+	if params.EmitterDomain == "" {
+		return &errortypes.BadInput{
+			Message: "Emitter domain undefined",
+		}
+	}
+
+	if params.MasterID == "" {
+		return &errortypes.BadInput{
+			Message: "MasterId undefined",
+		}
+	}
+
+	if params.SlaveID == "" {
+		return &errortypes.BadInput{
+			Message: "SlaveId undefined",
+		}
+	}
+
+	return nil
+}
+
+func (a *AdOceanAdapter) MakeBids(
+	internalRequest *openrtb.BidRequest,
+	externalRequest *adapters.RequestData,
+	response *adapters.ResponseData,
+) (*adapters.BidderResponse, []error) {
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Network error?", response.StatusCode)}
+	}
+
+	requestURL, _ := url.Parse(externalRequest.Uri)
+	queryParams := requestURL.Query()
+	auctionIDs := queryParams["aid"]
+
+	bidResponses := make([]ResponseAdUnit, 0)
+	if err := json.Unmarshal(response.Body, &bidResponses); err != nil {
+		return nil, []error{err}
+	}
+
+	var parsedResponses = adapters.NewBidderResponseWithBidsCapacity(len(auctionIDs))
+
+	for _, auctionFullID := range auctionIDs {
+		auctionIDsSlice := strings.SplitN(auctionFullID, ":", 2)
+		auctionID := auctionIDsSlice[1]
+		slaveID := auctionIDsSlice[0]
+
+		for _, bid := range bidResponses {
+			if bid.ID == slaveID {
+				if bid.Error == "true" {
+					break
+				}
+
+				price, _ := strconv.ParseFloat(bid.Price, 64)
+				width, _ := strconv.ParseUint(bid.Width, 10, 64)
+				height, _ := strconv.ParseUint(bid.Height, 10, 64)
+				adCode, err := prepareAdCodeForBid(bid)
+				if err != nil {
+					return nil, []error{err}
+				}
+
+				parsedResponses.Bids = append(parsedResponses.Bids, &adapters.TypedBid{
+					Bid: &openrtb.Bid{
+						ID:    bid.ID,
+						ImpID: auctionID,
+						Price: price,
+						AdM:   adCode,
+						CrID:  bid.CrID,
+						W:     width,
+						H:     height,
+					},
+					BidType: openrtb_ext.BidTypeBanner,
+				})
+				parsedResponses.Currency = bid.Currency
+
+				break
+			}
+		}
+	}
+
+	return parsedResponses, []error{}
+}
+
+func prepareAdCodeForBid(bid ResponseAdUnit) (string, error) {
+	sspCode, err := url.QueryUnescape(bid.Code)
+	if err != nil {
+		return "", err
+	}
+
+	measurementCode := `
+		<script>
+			+function() {
+				var wu = "%s";
+				var su = "%s".replace(/\[TIMESTAMP\]/, Date.now());
+
+				if (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) {
+					(new Image(1,1)).src = wu
+				}
+
+				if (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) {
+					(new Image(1,1)).src = su
+				}
+			}();
+		</script>
+	`
+	whiteSpace := regexp.MustCompile(`\s+`)
```

Comment:
I am thinking you may want to add this value as a parameter of your adapter object, and make this function a method of that object. Then you could compile `whitespace` just once on startup, rather than recompiling it every request.

Reply:
Thanks, it's changed. I've also added additional if to check if this object is not nil (on line 146), and corrected a little oversight at line 177 (instead of using newUri variable new string was generated).